### PR TITLE
Optimize multilocation reanchoring logic in XCM 

### PR DESCRIPTION
**Pull request by tonyalaribe**
This PR improves the algorithm for reanchor - see https://github.com/paritytech/polkadot/issues/4489 for more details.

Approach:
- We first normalize/prepend target and id with the ancestor
- We then remove any prefix nodes with both target and id share in common
- Then we convert all of target to parents, and add that to the parents of id. 
This gives us the path to go from target to id, without any superflous nodes, or the need for post simplification.

Question:
The `simplify()` method which was used previously is public. Does it make sense to leave it or should it be removed if it's not used anywhere else in the polkadot codebase (except tests)? This also applies with the `inverted()` method on Multilocation

### CODE REVIEW

**Review by KiChjang — APPROVED** [review]

**Comment by muharem** [comment]
@tonyalaribe I see two not covered edge cases, here the failing tests https://github.com/paritytech/polkadot/pull/6307

File: xcm/src/v1/multilocation.rs [hunk 1]
```diff
@@ -329,36 +329,21 @@ impl MultiLocation {
...
-		// 3. Given that we know some of `target` ancestry, ensure that any parents in `self` are
-		// strictly needed.
-		self.simplify(target.interior());
+		// Remove common parents between now normalized self and target
+		while target.interior().first() == self.interior().first() {
+			target.take_first_interior();
+			self.take_first_interior();
+		}
 
+		self.parents = self.parent_count() + target.parent_count() + target.interior().len() as u8;
```
Comment (muharem) [inline xcm/src/v1/multilocation.rs:343]: here the possibility of an overflow and self is modified.

File: xcm/src/v1/multilocation.rs [hunk 2]
```diff
@@ -329,36 +329,26 @@ impl MultiLocation {
...
+			target.take_first_interior();
+			id.take_first_interior();
+		}
 
-		// 3. Given that we know some of `target` ancestry, ensure that any parents in `self` are
-		// strictly needed.
-		self.simplify(target.interior());
+		id.parents = id
+			.parent_count()
+			.saturating_add(target.parent_count().saturating_add(target.interior().len() as u8));
```
Comment (muharem) [inline xcm/src/v1/multilocation.rs:346]:
```suggestion
		let final_parents = (id.parent_count() as usize).saturating_add(
			(target.parent_count() as usize).saturating_add(target.interior().len()),
		);
		if final_parents > 255 {
			return Err(()) 
		}
		id.parents = final_parents as u8;
```

File: xcm/src/v1/multilocation.rs [hunk 3]
```diff
@@ -903,6 +878,131 @@ mod tests {
...
+	}
+
+	#[test]
+	fn reanchor_overflow_handled() {
+		let mut id = MultiLocation::new(255, X1(Parachain(255)));
+		let original_id = id.clone();
+		let ancestry = (Parachain(500)).into();
+		let target = (Parachain(500), Parachain(501)).into();
+		// Does not overflow
+		id.reanchor(&target, &ancestry).unwrap();
```
Comment (muharem) [inline xcm/src/v1/multilocation.rs:1001]:
```suggestion
		id.reanchor(&target, &ancestry).err().unwrap();
```

File: xcm/src/v1/multilocation.rs [hunk 4]
```diff
@@ -903,6 +878,131 @@ mod tests {
...
+			(Parachain(2000), Parachain(2022), PalletInstance(50), GeneralIndex(32)).into();
+		let target = (Parent, Parachain(1000), Parachain(3000)).into();
+		let expected = (Parent, Parent, Parachain(4000), GeneralIndex(42)).into();
+		id.reanchor(&target, &ancestry).unwrap();
+		assert_eq!(id, expected);
+	}
+
+	#[test]
+	fn reanchor_overflow_handled() {
+		let mut id = MultiLocation::new(255, X1(Parachain(255)));
```
Comment (muharem) [inline xcm/src/v1/multilocation.rs:996]:
```suggestion
		let mut id = MultiLocation::new(254, X1(Parachain(255)));
```

Comment (muharem) [reply xcm/src/v1/multilocation.rs:996]: your test will fail with this change, your last assert pass now because when it overflows (to 256), it rounds it to 255.
with 254 it will be a better test.

File: xcm/src/v1/multilocation.rs (hunk 3 above)
Comment (muharem) [reply xcm/src/v1/multilocation.rs:1001]: when it overflows we expect the error

File: xcm/src/v1/multilocation.rs (hunk 2 above)
Comment (tonyalaribe) [reply xcm/src/v1/multilocation.rs]: Oh, I see. The idea is to return an error if there's an overflow and not just saturate it to fail silently? In that case I've switched to using a `checked_add` instead of `saturating_add`. That should give us the behaviour we want.

**Review by muharem — APPROVED** [review]

**Comment by tonyalaribe** [comment]
bot rebase

**Comment by tonyalaribe** [comment]
bot merge

**Comment by tonyalaribe** [comment]
bot rebase

**Comment by tonyalaribe** [comment]
bot rebase

**Comment by tonyalaribe** [comment]
bot rebase

**Comment by tonyalaribe** [comment]
bot rebase

**Comment by dzmitry-lahoda** [comment]
I took tests from `location_conversion` and ported to `multilocation` on `master` of polkadot
```
	#[test]
	fn inverter_works_in_tree() {
		let ancestry: MultiLocation = X3(Parachain(1), account20(), account20()).into();
		let target = MultiLocation::new(3, X2(Parachain(2), account32()));
		let inverted = ancestry.inverted(&target).unwrap();
		let expected = MultiLocation::new(2, X3(Parachain(1), account20(), account20()));
		assert_eq!(inverted, expected);
	}

	#[test]
	fn inverter_uses_ancestry_as_inverted_location() {
		let ancestry: MultiLocation = X2(account20(), account20()).into();
		let target = MultiLocation::grandparent();
		let inverted = ancestry.inverted(&target).unwrap();
		let expected = X2(account20(), account20()).into();
		assert_eq!(inverted, expected);
	}

	#[test]
	fn inverter_uses_only_child_on_missing_ancestry() {
		let ancestry: MultiLocation = X1(PalletInstance(5)).into();
		let target = MultiLocation::grandparent();
		let inverted = ancestry.inverted(&target).unwrap();
		let expected = X2(PalletInstance(5), OnlyChild).into();
		assert_eq!(inverted, expected);
	}

	#[test]
	fn inverter_errors_when_location_is_too_large() {
		let ancestry: MultiLocation = Here.into();
		let target = MultiLocation { parents: 99, interior: X1(Parachain(88)) };
		let inverted = ancestry.inverted(&target);
		assert_eq!(inverted, Err(()));
	}
```

Only one test is failing:
```
---- v1::multilocation::tests::inverter_uses_only_child_on_missing_ancestry stdout ----
thread 'v1::multilocation::tests::inverter_uses_only_child_on_missing_ancestry' panicked at 'assertion failed: `(left == right)`
  left: `MultiLocation { parents: 0, interior: X2(OnlyChild, PalletInstance(5)) }`,
 right: `MultiLocation { parents: 0, interior: X2(PalletInstance(5), OnlyChild) }`', xcm/src/v1/multilocation.rs:913:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

May be useful to ensure that no regression happened.
